File: IAIso-v5.0/core/iaiso-java/iaiso-conformance/src/main/java/io/iaiso/conformance/PolicyRunner.java
package io.iaiso.conformance;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import io.iaiso.policy.Policy;
import io.iaiso.policy.PolicyException;
import io.iaiso.policy.PolicyLoader;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

final class PolicyRunner {
    static final double TOLERANCE = 1e-9;

    private PolicyRunner() {}

    static List<VectorResult> run(Path specRoot) throws IOException {
        Path file = specRoot.resolve("policy").resolve("vectors.json");
        JsonObject doc = JsonParser.parseString(
            new String(Files.readAllBytes(file))).getAsJsonObject();
        List<VectorResult> out = new ArrayList<>();
        if (doc.has("valid")) {
            for (JsonElement el : doc.getAsJsonArray("valid")) {
                out.add(runValid(el.getAsJsonObject()));
            }
        }
        if (doc.has("invalid")) {
            for (JsonElement el : doc.getAsJsonArray("invalid")) {
                out.add(runInvalid(el.getAsJsonObject()));
            }
        }
        return out;
    }

    private static VectorResult runValid(JsonObject v) {
        String name = "valid/" + v.get("name").getAsString();
        try {
            Policy p = PolicyLoader.build(v.get("document"));

            if (v.has("expected_pressure") && v.get("expected_pressure").isJsonObject()) {
                JsonObject ep = v.getAsJsonObject("expected_pressure");
                String err = checkPressure(p, ep);
                if (err != null) return VectorResult.fail("policy", name, err);
            }

            if (v.has("expected_consent") && v.get("expected_consent").isJsonObject()) {
                JsonObject ec = v.getAsJsonObject("expected_consent");
                String err = checkConsent(p, ec);
                if (err != null) return VectorResult.fail("policy", name, err);
            }

            if (v.has("expected_metadata") && v.get("expected_metadata").isJsonObject()) {
                JsonObject em = v.getAsJsonObject("expected_metadata");
                if (em.size() != p.getMetadata().size()) {
                    return VectorResult.fail("policy", name,
                        "metadata size: got " + p.getMetadata().size()
                        + ", want " + em.size());
                }
            }

            return VectorResult.pass("policy", name);
        } catch (RuntimeException e) {
            return VectorResult.fail("policy", name, "build failed: " + e.getMessage());
        }
    }

    private static String checkPressure(Policy p, JsonObject ep) {
        double[][] checks = {
            {p.getPressure().getTokenCoefficient(),
                ep.has("token_coefficient") ? ep.get("token_coefficient").getAsDouble() : Double.NaN},
            {p.getPressure().getToolCoefficient(),
                ep.has("tool_coefficient") ? ep.get("tool_coefficient").getAsDouble() : Double.NaN},
            {p.getPressure().getDepthCoefficient(),
                ep.has("depth_coefficient") ? ep.get("depth_coefficient").getAsDouble() : Double.NaN},
            {p.getPressure().getDissipationPerStep(),
                ep.has("dissipation_per_step") ? ep.get("dissipation_per_step").getAsDouble() : Double.NaN},
            {p.getPressure().getDissipationPerSecond(),
                ep.has("dissipation_per_second") ? ep.get("dissipation_per_second").getAsDouble() : Double.NaN},
            {p.getPressure().getEscalationThreshold(),
                ep.has("escalation_threshold") ? ep.get("escalation_threshold").getAsDouble() : Double.NaN},
            {p.getPressure().getReleaseThreshold(),
                ep.has("release_threshold") ? ep.get("release_threshold").getAsDouble() : Double.NaN},
        };
        String[] labels = {
            "token_coefficient", "tool_coefficient", "depth_coefficient",
            "dissipation_per_step", "dissipation_per_second",
            "escalation_threshold", "release_threshold"
        };
        for (int i = 0; i < checks.length; i++) {
            if (Double.isNaN(checks[i][1])) continue;
            if (Math.abs(checks[i][0] - checks[i][1]) > TOLERANCE) {
                return labels[i] + ": got " + checks[i][0] + ", want " + checks[i][1];
            }
        }
        if (ep.has("post_release_lock")) {
            boolean want = ep.get("post_release_lock").getAsBoolean();
            if (want != p.getPressure().isPostReleaseLock()) {
                return "post_release_lock mismatch";
            }
        }
        return null;
    }

    private static String checkConsent(Policy p, JsonObject ec) {
        if (ec.has("issuer")) {
            String want = ec.get("issuer").isJsonNull() ? null : ec.get("issuer").getAsString();
            String got = p.getConsent().getIssuer();
            if (want == null ? got != null : !want.equals(got)) {
                return "consent.issuer: got " + got + ", want " + want;
            }
        }
        if (ec.has("default_ttl_seconds")) {
            double want = ec.get("default_ttl_seconds").getAsDouble();
            if (Math.abs(p.getConsent().getDefaultTtlSeconds() - want) > TOLERANCE) {
                return "default_ttl_seconds: got " + p.getConsent().getDefaultTtlSeconds()
                    + ", want " + want;
            }
        }
        if (ec.has("required_scopes")) {
            JsonArray want = ec.getAsJsonArray("required_scopes");
            if (want.size() != p.getConsent().getRequiredScopes().size()) {
                return "required_scopes length mismatch";
            }
        }
        if (ec.has("allowed_algorithms")) {
            JsonArray want = ec.getAsJsonArray("allowed_algorithms");
            if (want.size() != p.getConsent().getAllowedAlgorithms().size()) {
                return "allowed_algorithms length mismatch";
            }
        }
        return null;
    }

    private static VectorResult runInvalid(JsonObject v) {
        String name = "invalid/" + v.get("name").getAsString();
        String expectPath = v.get("expect_error_path").getAsString();
        try {
            // For non-mapping documents we still need to call build()
            JsonElement doc = v.get("document");
            PolicyLoader.build(doc);
            return VectorResult.fail("policy", name,
                "expected error containing '" + expectPath + "', got Ok");
        } catch (PolicyException e) {
            if (!e.getMessage().contains(expectPath)) {
                return VectorResult.fail("policy", name,
                    "expected error containing '" + expectPath + "', got: " + e.getMessage());
            }
            return VectorResult.pass("policy", name);
        }
    }
}

File: IAIso-v5.0/core/iaiso-rust/crates/conformance/src/policy_runner.rs
use crate::VectorResult;
use iaiso_policy::build;
use serde::Deserialize;
use serde_json::Value;
use std::path::Path;

#[derive(Debug, Deserialize)]
struct ValidPolicyVector {
    name: String,
    #[serde(default)]
    description: String,
    document: Value,
    #[serde(default)]
    expected_pressure: Option<Value>,
    #[serde(default)]
    expected_consent: Option<Value>,
    #[serde(default)]
    expected_metadata: Option<Value>,
}

#[derive(Debug, Deserialize)]
struct InvalidPolicyVector {
    name: String,
    #[serde(default)]
    description: String,
    document: Value,
    expect_error_path: String,
}

#[derive(Debug, Deserialize)]
struct PolicyFile {
    valid: Vec<ValidPolicyVector>,
    invalid: Vec<InvalidPolicyVector>,
}

pub fn run_policy(spec_root: &Path) -> Result<Vec<VectorResult>, String> {
    let path = spec_root.join("policy").join("vectors.json");
    let data = std::fs::read(&path).map_err(|e| format!("read {}: {}", path.display(), e))?;
    let file: PolicyFile =
        serde_json::from_slice(&data).map_err(|e| format!("parse {}: {}", path.display(), e))?;

    let mut results = Vec::new();
    for v in file.valid {
        results.push(run_one_valid(v));
    }
    for v in file.invalid {
        results.push(run_one_invalid(v));
    }
    Ok(results)
}

fn close(a: f64, b: f64) -> bool {
    (a - b).abs() <= crate::TOLERANCE
}

fn check_f64(label: &str, got: f64, want: Option<&Value>) -> Option<String> {
    if let Some(v) = want {
        if let Some(w) = v.as_f64() {
            if !close(got, w) {
                return Some(format!("{}: got {}, want {}", label, got, w));
            }
        }
    }
    None
}

fn run_one_valid(v: ValidPolicyVector) -> VectorResult {
    let mut r = VectorResult {
        section: "policy".to_string(),
        name: format!("valid/{}", v.name),
        passed: false,
        message: String::new(),
    };
    let p = match build(&v.document) {
        Ok(p) => p,
        Err(e) => {
            r.message = format!("build failed: {}", e);
            return r;
        }
    };

    if let Some(ep) = &v.expected_pressure {
        let map = match ep.as_object() {
            Some(m) => m,
            None => {
                r.message = "expected_pressure not an object".to_string();
                return r;
            }
        };
        for (label, got) in [
            ("token_coefficient", p.pressure.token_coefficient),
            ("tool_coefficient", p.pressure.tool_coefficient),
            ("depth_coefficient", p.pressure.depth_coefficient),
            ("dissipation_per_step", p.pressure.dissipation_per_step),
            ("dissipation_per_second", p.pressure.dissipation_per_second),
            ("escalation_threshold", p.pressure.escalation_threshold),
            ("release_threshold", p.pressure.release_threshold),
        ] {
            if let Some(msg) = check_f64(label, got, map.get(label)) {
                r.message = msg;
                return r;
            }
        }
        if let Some(want) = map.get("post_release_lock").and_then(Value::as_bool) {
            if p.pressure.post_release_lock != want {
                r.message = "post_release_lock mismatch".to_string();
                return r;
            }
        }
    }

    if let Some(ec) = &v.expected_consent {
        if let Some(map) = ec.as_object() {
            // issuer: vectors use null for "no issuer set"; our type uses Option<String>
            if let Some(want) = map.get("issuer") {
                let want_s = want.as_str();
                let got_s = p.consent.issuer.as_deref();
                if want_s != got_s {
                    r.message = format!(
                        "consent.issuer: got {:?}, want {:?}",
                        got_s, want_s
                    );
                    return r;
                }
            }
            if let Some(msg) = check_f64(
                "default_ttl_seconds",
                p.consent.default_ttl_seconds,
                map.get("default_ttl_seconds"),
            ) {
                r.message = msg;
                return r;
            }
            if let Some(want) = map.get("required_scopes").and_then(Value::as_array) {
                if want.len() != p.consent.required_scopes.len() {
                    r.message = "required_scopes length mismatch".to_string();
                    return r;
                }
            }
            if let Some(want) = map.get("allowed_algorithms").and_then(Value::as_array) {
                if want.len() != p.consent.allowed_algorithms.len() {
                    r.message = "allowed_algorithms length mismatch".to_string();
                    return r;
                }
            }
        }
    }

    if let Some(em) = &v.expected_metadata {
        if let Some(map) = em.as_object() {
            if map.len() != p.metadata.len() {
                r.message = format!(
                    "metadata size: got {}, want {}",
                    p.metadata.len(),
                    map.len()
                );
                return r;
            }
        }
    }

    r.passed = true;
    r
}

fn run_one_invalid(v: InvalidPolicyVector) -> VectorResult {
    let mut r = VectorResult {
        section: "policy".to_string(),
        name: format!("invalid/{}", v.name),
        passed: false,
        message: String::new(),
    };
    match build(&v.document) {
        Ok(_) => {
            r.message = format!(
                "expected error containing {:?}, got Ok",
                v.expect_error_path
            );
        }
        Err(e) => {
            let msg = e.to_string();
            if msg.contains(&v.expect_error_path) {
                r.passed = true;
            } else {
                r.message = format!(
                    "expected error containing {:?}, got {:?}",
                    v.expect_error_path, msg
                );
            }
        }
    }
    r
}
